### PR TITLE
fix: sync run with callback

### PR DESCRIPTION
**Pull request by jsimecek**
Fixes for task https://github.com/apify/apify-zapier-integration/issues/134

fix:
- better regex matching for Actor not found errors
- stable synchronous runs using Zapier callback + Apify ad-hoc webhooks
  - flow: 
    - [`z.generateCallbackUrl`](https://docs.zapier.com/integrations/build-cli/core#z-generatecallbackurl) pauses the step (user sees a `Delayed` step state) and generates a callback url
    - run start send to Apify with callback url in [`webhooks`](https://docs.apify.com/integrations/webhooks/ad-hoc-webhooks) field
    - Apify runs until it gets into a terminal state
    - Apify hits the callback url
    - Zapier step resumes
    - Zapier fetches run results
    - enrich run in Zapier
    - Zapier finishes step with run results
  - edge cases:
    - test runs for individual steps don't work with `z.generateCallbackUrl`, so that path uses a 25s limit instead and returns partial results up to that point if the run didn't finish (documented in helpText); entire Zap test work normally
  - info:
    - will hit generic Zapier timeout only if Apify fails to hit the callback url and the Zap step waits 30 days, so likely never
    - we can adjust when the run should timeout in Apify using the `timeout` field (I set it to max 1 hour unless set lower by user/Actor/task)
    - the run can't timeout in Zapier and continue to run in Apify

Testable on version `4.6.2`, before the implementation of code review edits on version `4.6.1`.

### CODE REVIEW

File: src/creates/actor_run.js
Comment (jsimecek) [inline src/creates/actor_run.js]: Q: Do we want to now recommend **Run Actor** and **Run task** to use synchronous mode by default rather than the async mode for the better UX?

File: src/creates/actor_run.js [hunk 1]
```diff
@@ -135,9 +142,22 @@ const runActor = async (z, bundle) => {
         }
     }
 
-    let { data: run } = await requestActorOrThrowNotFound(z, requestOpts, actorId);
-    if (runSync) run = await waitForRunToFinish(z.request, run.id, DEFAULT_RUN_WAIT_TIME_OUT_SECONDS, true);
+    // Calling z.generateCallbackUrl() is what pauses the Zap step, so it must not be called when running async.
+    if (runSync) {
+        requestOpts.params.timeout = Math.min(timeoutSecs || DEFAULT_SYNC_RUN_TIMEOUT_SECS, DEFAULT_SYNC_RUN_TIMEOUT_SECS);
```
Comment (jsimecek) [inline src/creates/actor_run.js:151]: Q: Is it fine to hard cap sync mode at 1 hour? Do we want it only as a fallback instead of `0` for infinite time, so the user can set a higher timeout than 1 hour if he wants to? Or do we want to get rid of the 1 hour cap entirely?

File: src/apify_helpers.js [hunk 2]
```diff
@@ -243,6 +244,46 @@ const getActorRun = async (z, bundle) => {
...
+    const runId = bundle.outputData?.id || bundle.cleanedRequest?.resource?.id;
+    if (!runId) {
+        throw new Error('The Apify run callback did not contain a run ID, so the run results could not be loaded.');
+    }
+
+    const { data: run } = await wrapRequestWithRetries(z.request, {
+        url: `${APIFY_API_ENDPOINTS.actorRuns}/${runId}`,
+    });
+
+    if (run.status === ACTOR_JOB_STATUSES.TIMED_OUT) {
```
Comment (drobnikj) [inline src/apify_helpers.js:273]: Question: Only TIMED-OUT is rejected here, so a FAILED or ABORTED run resumes as a successful step and the Zap carries on with a failed run.

File: src/creates/actor_run.js
Comment (JanHranicky) [reply src/creates/actor_run.js]: n8n has it defaultly set to `true`, would do it the same here

File: src/creates/actor_run.js (hunk 1 above)
Comment (JanHranicky) [reply src/creates/actor_run.js:151]: Imo it is fine to have it capped at 1 Hour. Synchronously running an Actor run for 1 hour is much better fit for the async version and if an Actor didn't finish in an hour a descriptive error should be thrown already redirecting the user to use the async variant.

**Review by JanHranicky — APPROVED** [review]
LGMT :)

**Review by drobnikj — CHANGES_REQUESTED** [review]
Looks good!

Two notes from testing; 

1. Test step - now when I choose Run Sync and run test step. The run is always returns without Dataset Items as there is no waiting time for dataset results pop up. This makes it harder for user to map the output of steps for following Zap step. Can we wait in test step some time to produce results (not necessary all) for user to easily maps next steps? 

Image

2. When the Actor/task fails the fail state will not propagate into Zap flow and the workflow run seems success but it does not.

Image

Image

File: src/apify_helpers.js (hunk 2 above)
Comment (jsimecek) [reply src/apify_helpers.js:273]: 1. Good idea. I have removed the `z.generateCallbackUrl` call from the test step code path and added a 25s limit instead. Now if the run does not finish within the 25s, it will return partial results it got gathered up to that point. You can test it on version `4.6.2`.
2. Yes both `FAILED` or `ABORTED` go through as a successful step, so the user can deal with the result himself instead of breaking the whole Zap. This is the same as the behavior on the currently published app version. I made the `TIMED-OUT` status throw to mimic the timeout error we threw as part of the Zap step code before, even though now it is a run status instead of part of the code.

File: src/creates/actor_run.js [hunk 3]
```diff
@@ -135,9 +143,30 @@ const runActor = async (z, bundle) => {
...
+        if (!isTestStep) requestOpts.params.webhooks = buildRunCallbackWebhookParam(z.generateCallbackUrl());
+    }
+
+    const { data: run } = await requestActorOrThrowNotFound(z, requestOpts, actorId);
+
+    if (runSync) {
+        // The step is paused here and finished by performResume once the run reaches a terminal status.
+        if (!isTestStep) return run;
```
Comment (drobnikj) [inline src/creates/actor_run.js:159]: Note: this returns the raw API run, skipping the `_.omit(run, OMIT_ACTOR_RUN_FIELDS)` that every other path applies, so `userId`, `stats`, `options` and `output` become the paused step's output data and get picked up as output fields. Same in `task_run.js` and `scrape_single_url.js` — worth omitting before returning.

File: src/apify_helpers.js (hunk 2 above)
Comment (drobnikj) [inline src/apify_helpers.js]: Question: the callback also fires on FAILED and ABORTED, but only TIMED_OUT is turned into an error, so a failed run resumes the step as a success and the Zap continues. Now that `runSync` defaults to yes, is that asymmetry intentional?

File: src/creates/task_run.js [hunk 4]
```diff
@@ -1,10 +1,33 @@
...
+        if (timeoutSecs === undefined || timeoutSecs === null) {
+            const { data: actor } = await wrapRequestWithRetries(z.request, {
+                url: `${APIFY_API_ENDPOINTS.actors}/${task.actId}`,
+            });
+            timeoutSecs = actor.defaultRunOptions?.timeoutSecs;
+        }
+
+        return Math.min(timeoutSecs || DEFAULT_SYNC_RUN_TIMEOUT_SECS, DEFAULT_SYNC_RUN_TIMEOUT_SECS);
+    } catch (err) {
+        return DEFAULT_SYNC_RUN_TIMEOUT_SECS;
```
Comment (drobnikj) [inline src/creates/task_run.js]: Note: falling back to the 1 hour cap on a failed lookup *widens* the timeout for a task configured with a shorter one, since the explicit `timeout` param overrides it. Omitting the param on failure would let the task's own timeout stand.

File: src/creates/actor_run.js [hunk 5]
```diff
@@ -135,9 +143,30 @@ const runActor = async (z, bundle) => {
...
+
+        const waitedRun = await waitForRunToFinish(z.request, run.id, getRemainingTestStepWaitSecs(stepStartedAt));
```
Comment (drobnikj) [inline src/creates/actor_run.js]: Suggestion: when the 25s budget runs out here the user gets a RUNNING run with partial `datasetItems` and nothing explaining why — `scrape_single_url.js` adds a `warning` field for exactly this case, worth mirroring.

**Review by drobnikj — APPROVED** [review]
Last note, to fix, but pre approving
